### ISSUE TITLE
What's the fate of FixTypeAbstractions?

### ISSUE THREAD

**Issue by frank-emrich** [links-lang/links#548]
There is a transformation called `FixTypeAbstractions`, which is included in `frontend.ml`, but commented out by 1b72bf9d80fa57fb3712e6917c7f712d4734678d by @dhil . Was this on purpose? Do we know that disabling it doesn't break anything?

**Comment by dhil** [comment]
You can try enabling it. As I recall, it produces unsound code. I'd just delete it.

**Comment by jamescheney** [comment]
We seem to have been getting along perfectly well without it for the last 1.5 years.

I tend to think that if this module was doing something worthwhile to fix a deficiency in the result fo `typeSugar` then the right thing to do is rationally reconstruct `typeSugar` and reimplement it so that it  does that thing right the first time and does not require a second `fixing` pass.  Or, if the thing that `fixTypeAbstractions` was doing is not possible to do in the same pass as `typeSugar` then we need to revisit this.  But I think deleting for now is quite safe since it is now dead code.

@slindley do you disagree?  According to git blame you are the only person to have touched this code, once 8 years ago and a second time 2 years ago to cope with some OCaml language changes it looks like.

**Comment by frank-emrich** [comment]
> We seem to have been getting along perfectly well without it for the last 1.5 years.

I guess we wouldn't have noticed if removing it breaks things. It looks like this just patches up quantifiers *after* frontend type-checking. This means that the fixes (or lack thereof) would only be noticed when type-checking the IR.

But I guess the right approach is to remove it and move towards a situation where `typeSugar` returns sugar AST that translates to type-correct IR code, without relying on a separate fixing pass.

**Comment by slindley** [comment]
Yes, I disagree. The comments at the top describe features that are important for supporting inference with first-class polymorphism. Let's not delete it until I've paged back in what's going on and we can determine whether we need it in the pipeline or whether the functionality it performs can be carried out earlier on by `typeSugar`.

It does seem to be assuming uniqueness of names for type variables, an invariant which may be being violated e.g. by the effect handler implementation.

Ideally we should be able to add some test cases to exercise it.

**Comment by jamescheney** [comment]
So I tried enabling `FixTypeAbstractions` pass and running all tests and many fail - mostly to do with handlers, but a few others, such as 

```
Instantiate.ArityMismatch #132 (RESOLVED)
Flexible quantifier elimination (fix introduced in 514187b)
examples/handlers/transaction.links
examples/initialise-list.links
```
This suggests that we should have a look at #132.  It was opened January 2017, closed by #133 in January 2018 in the effects+handlers branch and merged into `master` in September 2017. and 514187b (October 2017)  In #133 @dhil mentions that `FixTypeAbstractions` had been commented out on August 1, 2015 but looking at the history/blame for `frontend.ml` I cannot find this change.

**Comment by dhil** [comment]
> [..] August 1, 2015 but looking at the history/blame for frontend.ml I cannot find this change.

I am pretty sure that @slindley was the one who disabled `FixTypeAbstractions`, but the change was made on my computer and subsequently pushed with my credentials, hence I became the author of the change. Moreover, right about August 1, 2015 I was about to finish my master's project (the initial implementation of effect handlers), which was when I first stumbled upon this issue.

> [..] but looking at the history/blame for frontend.ml I cannot find this change.

I think some of the history got lost back when we merged `sessions` and `master`, and subsequently `effects+sessions` branch. I seem to remember those branches had diverged quite significantly.

**Comment by jamescheney** [comment]
This appears to be closed by #662, which deletes `fixTypeAbstractions`.

Before closing, maybe @slindley and/or @dhil should confirm that the recent changes to the type system align with what algebraic effects/handlers needs - I would hope so.

**Comment by slindley** [comment]
The ordered quantification and FreezeML changes make `fixTypeAbstractions` redundant now.
